### PR TITLE
Update SonataUserBundle to 4.x

### PR DESCRIPTION
**Pull request by wbloszyk**
- change contribute SonataUserBundle repository to SonataUserBundle 4.x
- add missing sonata.user.block.accout as sonata.demo.block.account in DemoBundle
- extends FOSUserBundle to override templates

### CODE REVIEW

File: app/config/sonata/sonata_customer.yml [hunk 1]
```diff
@@ -1,5 +1,59 @@
...
+                domain: 'SonataCustomerBundle'
+                route_parameters:  {}
+            -
+                route:  'sonata_order_index'
+                label:  'order_list'
+                domain: 'SonataOrderBundle'
+                route_parameters:  {}
+
+        blocks:
+            #- { position: left, type: sonata.block.service.text, settings: { content: "<h2>Welcome!</h2> <p>This is a sample user profile dashboard, feel free to override it in the configuration! Want to make this text dynamic? For instance display the user's name? Create a dedicated block and edit the configuration!</p>"} }
```
Comment (core23) [inline app/config/sonata/sonata_customer.yml]: Please remove this

File: composer.json [hunk 2]
```diff
@@ -52,15 +45,15 @@
...
         "sonata-project/page-bundle": "^3.7.1",
         "sonata-project/formatter-bundle": "^3.4.1",
         "sonata-project/news-bundle": "^3.3.0",
         "sonata-project/exporter": "^1.8.0",
         "sonata-project/timeline-bundle": "^3.3.0",
         "sonata-project/classification-bundle": "^3.6.1",
-        "sonata-project/ecommerce": "^3.0",
+        "sonata-project/ecommerce": "^3.2",
         "sonata-project/comment-bundle": "^3.1.0",
         "sonata-project/google-authenticator": "^2.0",
```
Comment (core23) [inline composer.json]: Can you please normalize the `composer.json` file

File: src/AppBundle/Entity/Classification/Collection.php [hunk 3]
```diff
@@ -29,6 +29,7 @@ class Collection extends BaseCollection
      * @var int
      */
     protected $id;
+    protected $productCollection;
```
Comment (core23) [inline src/AppBundle/Entity/Classification/Collection.php]: PHPDoc missing

File: src/AppBundle/Entity/User/User.php [hunk 4]
```diff
@@ -30,6 +31,11 @@ class User extends BaseUser
      */
     protected $id;
 
+    /**
+     * @var ArrayCollection
```
Comment (core23) [inline src/AppBundle/Entity/User/User.php]:
```suggestion
     * @var Collection|CustomerInterface[]
```

File: src/Application/FOS/UserBundle/ApplicationFOSUserBundle.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * This file has been generated by the SonataEasyExtendsBundle.
+ *
+ * @see https://sonata-project.org/easy-extends
+ *
+ * References:
+ * @see http://symfony.com/doc/current/book/bundles.html
+ */
+class ApplicationFOSUserBundle extends Bundle
```
Comment (core23) [inline src/Application/FOS/UserBundle/ApplicationFOSUserBundle.php]:
```suggestion
final class ApplicationFOSUserBundle extends Bundle
```

File: src/Application/FOS/UserBundle/ApplicationFOSUserBundle.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public function getParent()
+    {
+        return 'FOSUserBundle';
```
Comment (core23) [inline src/Application/FOS/UserBundle/ApplicationFOSUserBundle.php]: This feature is deprecated in symfony 3+

Comment (wbloszyk) [reply src/Application/FOS/UserBundle/ApplicationFOSUserBundle.php]: ApplicationFOSUserBundle is used only for extends templates. In Symfony 4+ templates should be remove to `templates/FOSUserBundle`.

File: app/config/sonata/sonata_customer.yml (hunk 1 above)
Comment (wbloszyk) [reply app/config/sonata/sonata_customer.yml]:
```suggestion
```

File: composer.json (hunk 2 above)
Comment (wbloszyk) [reply composer.json]: I will made it in project update PR

File: src/AppBundle/Entity/Classification/Collection.php (hunk 3 above)
Comment (wbloszyk) [reply src/AppBundle/Entity/Classification/Collection.php]: done

File: src/AppBundle/Entity/Classification/Collection.php [hunk 7]
```diff
@@ -30,6 +32,11 @@ class Collection extends BaseCollection
      */
     protected $id;
 
+    /**
+     * @var DoctrineCollection|ProductCollectionInterface[]
```
Comment (phansys) [inline src/AppBundle/Entity/Classification/Collection.php]: So, can it hold a collection of objects implementing `ProductCollectionInterface`?

File: src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig [hunk 8]
```diff
@@ -0,0 +1,81 @@
+{#
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig]: Please, be consistent with the file headers.

File: src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                            </form>
+                        {% endblock %}
+                    </div>
+                </div>
+            {% endblock %}
+        </div>
+        <div class="col-sm-6">
+
+            {% block sonata_user_registration %}
+                {{ render(controller("fos_user.registration.controller:registerAction")) }}
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig]:
```suggestion
                {{ render(controller('fos_user.registration.controller:registerAction')) }}
```

File: src/Application/FOS/UserBundle/Resources/views/Profile/edit.html.twig [hunk 10]
```diff
@@ -0,0 +1,9 @@
+{% extends "@FOSUser/layout.html.twig" %}
+
+{% block sonata_profile_title %}
+    {{ "title_user_account" | trans({}, 'SonataUserBundle')}} - {{ "title_user_edit_profile" | trans({}, 'SonataUserBundle')}}
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/Profile/edit.html.twig]:
```suggestion
    {{ 'title_user_account' | trans({}, 'SonataUserBundle')}} - {{ 'title_user_edit_profile' | trans({}, 'SonataUserBundle')}}
```

File: src/Application/FOS/UserBundle/Resources/views/ChangePassword/change_password_content.html.twig [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{% trans_default_domain 'FOSUserBundle' %}
+
+<div class="panel panel-default">
+    <div class="panel-heading">
+        <h3 class="panel-title">{{ "title_user_edit_password" | trans({}, 'SonataUserBundle') }}</h3>
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/ChangePassword/change_password_content.html.twig]:
```suggestion
        <h3 class="panel-title">{{ 'title_user_edit_password' | trans({}, 'SonataUserBundle') }}</h3>
```

File: src/Application/FOS/UserBundle/Resources/views/ChangePassword/change_password.html.twig [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{% extends "@FOSUser/layout.html.twig" %}
+
+{% block sonata_profile_title %}
+    {{ "title_user_account" | trans({}, 'SonataUserBundle')}} - {{ "title_user_edit_password" | trans({}, 'SonataUserBundle') }}
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/ChangePassword/change_password.html.twig]:
```suggestion
    {{ 'title_user_account' | trans({}, 'SonataUserBundle')}} - {{ 'title_user_edit_password' | trans({}, 'SonataUserBundle') }}
```

File: app/config/sonata/sonata_customer.yml [hunk 13]
```diff
@@ -1,5 +1,58 @@
 # Enable Doctrine to map the provided entities
 sonata_customer:
+    profile:
+        template:       'SonataCustomerBundle:Profile:action.html.twig'
```
Comment (phansys) [inline app/config/sonata/sonata_customer.yml]: IMO, we could unalign the values in order to make easier future updates.

Comment (wbloszyk) [reply app/config/sonata/sonata_customer.yml]: I comment it, becouse this is default value

File: src/AppBundle/Entity/Classification/Collection.php (hunk 7 above)
Comment (wbloszyk) [reply src/AppBundle/Entity/Classification/Collection.php]: Look at:
https://github.com/sonata-project/ecommerce/pull/653/files

Comment (phansys) [reply src/AppBundle/Entity/Classification/Collection.php]: IIUC, this should be a "single level" collection (`ProductCollectionInterface`), not a collection of collections (`ProductCollectionInterface[]`).

Comment (wbloszyk) [reply src/AppBundle/Entity/Classification/Collection.php]: Should not be `productCollections`???

Look at:
```php
  $collector->addAssociation($config['class']['product'], 'mapOneToMany', [
             'fieldName' => 'productCollections',
             'targetEntity' => $config['class']['product_collection'],
             'cascade' => [
                'persist',
             ],
             'mappedBy' => 'product',
             'orphanRemoval' => false,
        ]);



$collector->addAssociation($config['class']['collection'], 'mapOneToMany', [
                'fieldName' => 'productCollection',
                'targetEntity' => $config['class']['product_collection'],
                'cascade' => [
                    'persist',
                ],
                'mappedBy' => 'collection',
                'orphanRemoval' => false,
            ]);
```

Comment (phansys) [reply src/AppBundle/Entity/Classification/Collection.php]: I'm not a "sonata-project/ecommerce" user, so I'm not familiarized with the collector concept.
If this entity is able to keep an association with many groups of product collections, I think the provided docblock is correct.

File: src/Sonata/Bundle/DemoBundle/Resources/views/Block/account.html.twig [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+#}
+
+{% extends sonata_block.templates.block_base %}
+
+{% block block %}
+    <div style="display:inline;">
+        {% if user %}
+            <a href="{{ path('fos_user_profile_show') }}">{{ user.username }}</a> |
+            <a href="{{ path('fos_user_security_logout') }}">{{ 'link_logout' | trans({}, "SonataUserBundle") }}</a>
```
Comment (phansys) [inline src/Sonata/Bundle/DemoBundle/Resources/views/Block/account.html.twig]:
```suggestion
            <a href="{{ path('fos_user_security_logout') }}">{{ 'link_logout' | trans({}, 'SonataUserBundle') }}</a>
```

**Comment by wbloszyk** [comment]
@core23 @phansys Ready to review. I remove productCollection and customers part from this PR. It should be in another atomic PR.

File: src/AppBundle/Entity/Classification/Collection.php (hunk 7 above)
Comment (jeroendk) [reply src/AppBundle/Entity/Classification/Collection.php]: @wbloszyk A Collection can have multiple ProductCollections as you stated. However, IMO you should not at this property to the Generic Collection Entity to keep the Entity clean. Therefore my PR: https://github.com/sonata-project/ecommerce/pull/653 so you won't be forced to.

File: app/config/sonata/sonata_block.yml [hunk 15]
```diff
@@ -55,23 +55,19 @@ sonata_block:
...
-        sonata.user.block.account:
-            cache: sonata.page.cache.js_async
+        # ProfileBundle's Blocks
+        sonata.demo.block.account:
+        #    cache: sonata.page.cache.js_async
             #cache: sonata.page.cache.esi
-            contexts: [user]
-        sonata.user.block.menu:
-            contexts: [user]
+         #   contexts: [profile]
```
Comment (phansys) [inline app/config/sonata/sonata_block.yml]: Please, fix indentation and try to be consistent with the existing comments.

File: app/config/sonata/sonata_customer.yml [hunk 16]
```diff
@@ -1,5 +1,57 @@
...
 sonata_customer:
+    profile:
+        #template:       '@SonataCustomer/Profile/action.html.twig'
+        #menu_builder:   'sonata.customer.profile.menu_builder.default'
+        menu:
+            -
+                route: 'sonata_customer_dashboard'
+                label: 'link_list_dashboard'
+                domain: 'SonataCustomerBundle'
+                route_parameters:  {}
```
Comment (phansys) [inline app/config/sonata/sonata_customer.yml]:
```suggestion
                route_parameters: {}
```

File: app/config/sonata/sonata_customer.yml [hunk 17]
```diff
@@ -1,5 +1,57 @@
 # Enable Doctrine to map the provided entities
 sonata_customer:
+    profile:
+        #template:       '@SonataCustomer/Profile/action.html.twig'
+        #menu_builder:   'sonata.customer.profile.menu_builder.default'
```
Comment (phansys) [inline app/config/sonata/sonata_customer.yml]:
```suggestion
        #template: '@SonataCustomer/Profile/action.html.twig'
        #menu_builder: 'sonata.customer.profile.menu_builder.default'
```

File: src/Application/FOS/UserBundle/Resources/views/Registration/register_content.html.twig [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+{% trans_default_domain 'FOSUserBundle' %}
+
+<div class="panel panel-success">
+    <div class="panel-heading">
+        <h2 class="panel-title">{{ 'title_user_registration'|trans({}, 'SonataUserBundle') }}</h2>
+    </div>
+    <div class="panel-body">
+        {{ form_start(form, {'method': 'post', 'action': path('fos_user_registration_register'), 'attr': {'class': 'fos_user_registration_register form-horizontal'}}) }}
+        {#
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/Registration/register_content.html.twig]: Why is this comment required?

File: src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+                            </form>
+                        {% endblock %}
+                    </div>
+                </div>
+            {% endblock %}
+        </div>
+        <div class="col-sm-6">
+
+            {% block sonata_user_registration %}
+                {{ render(controller('fos_user.registration.controller:registerAction')) }}
```
Comment (phansys) [inline src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig]: I'm not sure, but this should not be written as follows?
```suggestion
                {{ render(controller('fos_user.registration.controller::registerAction')) }}
```

Comment (wbloszyk) [reply src/Application/FOS/UserBundle/Resources/views/Security/base_login.html.twig]: It is correct now. When I add `::` then it make error.

File: src/Application/FOS/UserBundle/Resources/views/Registration/register_content.html.twig (hunk 18 above)
Comment (wbloszyk) [reply src/Application/FOS/UserBundle/Resources/views/Registration/register_content.html.twig]: form_enctype is deprecated, I will remove this comment.

File: app/config/sonata/sonata_customer.yml (hunk 17 above)
Comment (wbloszyk) [reply app/config/sonata/sonata_customer.yml]: All configs have indentation like that. It is more readable.

File: app/config/sonata/sonata_customer.yml (hunk 16 above)
Comment (wbloszyk) [reply app/config/sonata/sonata_customer.yml]: Here change is good :)

File: app/config/sonata/sonata_customer.yml (hunk 17 above)
Comment (phansys) [reply app/config/sonata/sonata_customer.yml]: I think we don't have a CS rule for this, so I'll ignore my suggestion.
Personally, I think alignments like this make harder the merge process when some of these lines are added or updated.

**Comment by wbloszyk** [comment]
@core23 @phansys Should I do someting here or it is ready to merge?

I have prepared next RP

**Comment by wbloszyk** [comment]
@core23 @phansys @OskarStark @greg0ire Update
reference to #616

**Comment by wbloszyk** [comment]
@core23 @phansys @greg0ire It is RTM. can you check it?

I have ready sandbox in Symfony 4, i will update it in some steps:
- fix project structure and configuration files
- Update to Symfony 4 and flex

File: app/config/sonata/sonata_block.yml [hunk 20]
```diff
@@ -55,23 +55,20 @@ sonata_block:
...
-            cache: sonata.page.cache.js_async
-            #cache: sonata.page.cache.esi
-            contexts: [user]
-        sonata.user.block.menu:
-            contexts: [user]
+        sonata.demo.block.account:
+            # cache: sonata.page.cache.js_async
+            # cache: sonata.page.cache.esi
+            # contexts: [profile]
+
```
Comment (phansys) [inline app/config/sonata/sonata_block.yml:71]:
```suggestion
```

**Comment by greg0ire** [comment]
Thanks @wbloszyk !

**Pull request merged**